### PR TITLE
Validate Extrinsic before it is gossiped or applied

### PR DESCRIPTION
**Pull request by aang114**
This PR does the following 2 things:

- Ensures that the extrinsic does not contain an invalid call or does not exceed maximum depth length **before it is gossiped**. This is done in `<Runtime as TaggedTransactionQueue<Block>>::validate_transaction()`
- Ensures that the extrinsic does not contain an invalid call or does not exceed maximum depth length **before it is applied**. This is done in `<Runtime as sp_block_builder::BlockBuilder<Block>>::apply_extrinsic()`

P.S: This PR also limits the maximum length (in bytes) that the metadata data of a Proto-Fragment / Fragment Definition / Fragment Instance can be

### CODE REVIEW

File: runtime/src/lib.rs
Comment (aang114) [inline runtime/src/lib.rs]: @sinkingsugar - seems it is possible access a pallet's storage item in `validate_transaction()` 🤔. 

(this function `is_valid()` is called by `<Runtime as TaggedTransactionQueue<Block>::validate_transaction()` btw)

File: runtime/src/lib.rs
Comment (aang114) [inline runtime/src/lib.rs]: @sinkingsugar - how exactly do I verify if this shards script is implementing all the traits it specified in `Categories::Shards::implementing` 😅 ? Thanks :)

File: runtime/src/lib.rs
Comment (aang114) [inline runtime/src/lib.rs]: @sinkingsugar - what do we do about all file types whose file signature / magic number (https://en.wikipedia.org/wiki/List_of_file_signatures) **does not exist in the `infer` crate?**

Should I manually verify them (i.e search online for their magic number and verify it against `data`) or just leave them as `false` - since anyways this is literally just a band-aid solution that won't be used for the real thing 😅?

(Btw all the file types that can't be verified using `infer` crate are labelled as `// TODO - Giovanni Petrantoni` 😅 )

File: runtime/src/lib.rs
Comment (aang114) [inline runtime/src/lib.rs]: @sinkingsugar  -  no idea what the file extension for this is even supposed to be 😅

File: runtime/src/lib.rs
Comment (aang114) [inline runtime/src/lib.rs]: this is only a partial solution btw since we're not checking for the file extension `.sdf`. Is that okay ?

File: runtime/src/lib.rs
Comment (sinkingsugar) [reply runtime/src/lib.rs]: Don't need to put my name lol

File: runtime/src/lib.rs
Comment (sinkingsugar) [reply runtime/src/lib.rs]: Our format

File: runtime/src/lib.rs
Comment (sinkingsugar) [reply runtime/src/lib.rs]: sdf is our format, not sqlite, set false

File: runtime/src/lib.rs
Comment (sinkingsugar) [reply runtime/src/lib.rs]: you don't for now it's fine

**Comment by aang114** [comment]
This PR does the following things:
1. Ensures that if the `data: Vec<u8>` param of certain calls is incorrect (see `does_call_have_correct_data_param()` for check logic)
   - an extrinsic with this call is never gossiped (ensured using `validate_transaction()`) 
   - the call is never applied (ensured using `<BaseCallFilter as Contains>::contains()`). 
2. Ensures that an extrinsic with a call that  has a depth level greater than 4 (see  `does_call_exceed_maximum_depth_level()` for check logic)
   -  it is never applied (ensured in `apply_extrinsic()`) and 
   - it is never gossiped (ensured in `validate_transaction()`)
3. Ensures that if a batch call contains uses `transaction_index::index`, it is never applied.

**Comment by aang114** [comment]
Hi @sinkingsugar, please approve this PR (if okay) so that it can be merged. Thanks

File: runtime/src/lib.rs
Comment (sinkingsugar) [inline runtime/src/lib.rs]: true, let it pass

File: runtime/src/lib.rs
Comment (sinkingsugar) [inline runtime/src/lib.rs]: true, let it pass for now

File: runtime/src/lib.rs
Comment (sinkingsugar) [inline runtime/src/lib.rs:467]: `.glb` cannot?

File: runtime/src/lib.rs
Comment (sinkingsugar) [inline runtime/src/lib.rs]: for now do `infer::is(data, "wasm")`

File: runtime/src/lib.rs
Comment (sinkingsugar) [inline runtime/src/lib.rs]: true

File: runtime/src/lib.rs
Comment (sinkingsugar) [inline runtime/src/lib.rs]: true

**Review by sinkingsugar — CHANGES_REQUESTED** [review]

**Comment by sinkingsugar** [comment]
@aang114 
I have a request before merging also:
We need to enforce a few categories to be `Local` only.
Those are:
- [ ] Shards
- [ ] Traits
- [ ] Text

**Comment by aang114** [comment]
> @aang114 I have a request before merging also: We need to enforce a few categories to be `Local` only. Those are:
> 
> * [ ]  Shards
> * [ ]  Traits
> * [ ]  Text

I'm not sure what you mean? What does `Local` mean?

File: runtime/src/lib.rs
Comment (aang114) [reply runtime/src/lib.rs:467]: That file extension doesn't exist in the `infer` crate

Comment (aang114) [reply runtime/src/lib.rs:467]: > That file extension doesn't exist in the `infer` crate

so I'll just put `true` for now

**Comment by aang114** [comment]
> @aang114 I have a request before merging also: We need to enforce a few categories to be `Local` only. Those are:
> 
> * [ ]  Shards
> * [ ]  Traits
> * [ ]  Text

done @sinkingsugar

File: runtime/src/lib.rs
Comment (sinkingsugar) [inline runtime/src/lib.rs]: let's do false for now

File: runtime/src/lib.rs
Comment (sinkingsugar) [inline runtime/src/lib.rs]: this makes this whole PR worthless, let's do false 👍

**Review by sinkingsugar — CHANGES_REQUESTED** [review]
just some final things

File: runtime/src/lib.rs
Comment (aang114) [reply runtime/src/lib.rs]: for everything I just made `true`, right?

Comment (sinkingsugar) [reply runtime/src/lib.rs]: what do you mean?
I mean a malicious party can select gltf and upload anything.

Comment (aang114) [reply runtime/src/lib.rs]: okay nevermind. I thought you wanted me to revert all the comments earlier where you said "put true". So I'll just make this `false`

Comment (aang114) [reply runtime/src/lib.rs]: done

**Review by sinkingsugar — APPROVED** [review]
CI green and do a big formatting pass cargo fmt or so before merge please

**Pull request merged**
